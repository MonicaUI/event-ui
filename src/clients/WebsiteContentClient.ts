import { RequestBuilder, fetchAndRetryIfServerBusy } from '../utils/wrappedFetchAndRetryIfServerTooBusy';
import { SESSION_HEADER_EVENT_ID } from '@cvent/event-ui-networking';
import { ServiceError } from '@cvent/event-ui-networking';

async function fetchOrThrow(request, errorMessage) {
  const response = await fetchAndRetryIfServerBusy(request);
  if (!response.ok) {
    throw await ServiceError.create(errorMessage, response, request);
  }
  if (response.status === 204) {
    return;
  }
  return await response.json();
}

/**
 * The client to fetch website related content (including archive page data) for the event.
 */
export default class WebsiteContentClient {
  baseUrl: $TSFixMe;
  environment: $TSFixMe;
  eventId: $TSFixMe;
  constructor(baseUrl?: $TSFixMe, eventId?: $TSFixMe, environment?: $TSFixMe) {
    this.baseUrl = baseUrl;
    this.eventId = eventId;
    this.environment = environment;
  }
  _getRequestBuilder(urlPath = ''): $TSFixMe {
    const builder = new RequestBuilder({ url: `${this.baseUrl}${urlPath}` })
      .header(SESSION_HEADER_EVENT_ID, this.eventId)
      .withCookies();
    return this.environment ? builder.query('environment', this.environment) : builder;
  }

  /**
   * Retrieves the archived page data for the event.
   * @param eventId - eventId of the event we want to fetch archive page data for
   * @param version Indicates which version of the website to use when fetching the archived page data.
   */
  async getEventArchivePageData(eventId: $TSFixMe, version: $TSFixMe): Promise<$TSFixMe> {
    const requestBuilder = this._getRequestBuilder(`websiteContent/${eventId}/eventArchivePage`)
      .query('snapshotVersion', version)
      .get();
    const request = requestBuilder.build();
    return await fetchOrThrow(request, `Failed to fetch archived page data for eventId: ${String(eventId)}`);
  }
}
