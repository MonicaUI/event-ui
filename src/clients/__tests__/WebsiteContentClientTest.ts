import WebsiteContentClient from '../WebsiteContentClient';
import defaultArchivePageData from '../../archivePage/fixtures/defaultArchivePageData.json';

jest.mock('@cvent/nucleus-networking', () => {
  const RequestBuilderModule = jest.requireActual<$TSFixMe>('@cvent/nucleus-networking');
  class MockRequestBuilder extends RequestBuilderModule.RequestBuilder {
    constructor(requestOptions = {}) {
      const newRequestOptions = {
        ...requestOptions,
        headers: {
          ...(requestOptions as $TSFixMe).headers,
          httplogpageloadid: 'httplogpageloadid',
          httplogrequestid: 'httplogrequestid'
        }
      };
      super(newRequestOptions);
    }
  }
  let mockResponse = {};
  let mockHeaders = {};
  return {
    ...RequestBuilderModule,
    RequestBuilder: MockRequestBuilder,
    setMockResponse(response, headers) {
      mockResponse = response;
      mockHeaders = headers;
    },
    fetchAndRetryIfServerBusy: jest.fn(() => {
      return Promise.resolve({
        ok: true,
        headers: {
          get(header) {
            return mockHeaders[header];
          },
          has(header) {
            return !!mockHeaders[header];
          }
        },
        json: () => Promise.resolve(mockResponse)
      });
    })
  };
});

// eslint-disable-next-line @typescript-eslint/no-var-requires
const { setMockResponse } = require('@cvent/nucleus-networking');

const client = new WebsiteContentClient('https://example.com/websiteContent/', 'eventId', 'dev');

beforeEach(() => {
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  require('@cvent/nucleus-networking').fetchAndRetryIfServerBusy.mockClear();
});

test('WebsiteContentClient.getEventArchivePageData matches snapshot', async () => {
  setMockResponse(defaultArchivePageData, {});
  const response = await client.getEventArchivePageData('eventId', {
    version: 'eventSnapshotVersion',
    registrationTypeId: 'registrationTypeId'
  });
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  const fetch = require('@cvent/nucleus-networking').fetchAndRetryIfServerBusy;
  expect(fetch).toHaveBeenCalled();
  expect(fetch.mock.calls[0]).toMatchSnapshot();

  // ensure response is unmodified when returned by client method
  expect(response).toEqual(defaultArchivePageData);
});
