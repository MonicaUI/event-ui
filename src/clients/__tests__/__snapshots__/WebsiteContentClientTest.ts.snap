// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WebsiteContentClient.getEventArchivePageData matches snapshot 1`] = `
Array [
  Request {
    "_bodyInit": undefined,
    "_bodyText": "",
    "credentials": "include",
    "headers": Headers {
      "map": Object {
        "cvent-event-id": "eventId",
        "httplogpageloadid": "httplogpageloadid",
        "httplogrequestid": "httplogrequestid",
      },
    },
    "method": "GET",
    "mode": null,
    "referrer": null,
    "url": "https://example.com/websiteContent/websiteContent/eventId/eventArchivePage?environment=dev&snapshotVersion=",
  },
]
`;
